monai/data/wsi_reader.py

```diff
@@ -805,106 +1025,125 @@ class TiffFileWSIReader(BaseWSIReader):
+    def get_downsample_ratio(self, wsi, level: int) -> float:
         """
         Returns the down-sampling ratio of the whole slide image at a given level.
 
         Args:
-            wsi: a whole slide image object loaded from a file
-            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
-                will be used.
+            wsi: a whole slide image object loaded from a file.
+            level: the level number where the downsample ratio is calculated.
 
         """
-        if level is None:
-            level = self.level
-
         return float(wsi.pages[0].imagelength) / float(wsi.pages[level].imagelength)
 
     @staticmethod
     def get_file_path(wsi) -> str:
         """Return the file path for the WSI object"""
         return str(abspath(wsi.filehandle.path))
 
-    def get_mpp(self, wsi, level: int | None = None) -> tuple[float, float]:
+    def get_mpp(self, wsi, level: int) -> tuple[float, float]:
         """
         Returns the micro-per-pixel resolution of the whole slide image at a given level.
 
         Args:
-            wsi: a whole slide image object loaded from a file
-            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
-                will be used.
+            wsi: a whole slide image object loaded from a file.
+            level: the level number where the mpp is calculated.
+
+        """
+        if (
+            "XResolution" in wsi.pages[level].tags
+            and "YResolution" in wsi.pages[level].tags
+            and wsi.pages[level].tags["XResolution"].value
+            and wsi.pages[level].tags["YResolution"].value
+        ):
+            unit = wsi.pages[level].tags.get("ResolutionUnit")
+            if unit is not None:
+                unit = unit.value
+                if unit == unit.CENTIMETER:
+                    factor = 10000.0
+                elif unit == unit.MILLIMETER:
+                    factor = 1000.0
+                elif unit == unit.MICROMETER:
+                    factor = 1.0
+                elif unit == unit.INCH:
+                    factor = 25400.0
+                else:
+                    warnings.warn(
+                        f"The resolution unit is not a valid tiff resolution, unit={unit}."
+                        " `micrometer` will be used as default."
+                    )
+                    factor = 1.0
+            else:
+                warnings.warn("The resolution unit is missing. `micrometer` will be used as default.")
+                factor = 1.0
 
+            # Here x and y resolutions are rational numbers so each of them is represented by a tuple.
+            yres = wsi.pages[level].tags["YResolution"].value
+            xres = wsi.pages[level].tags["XResolution"].value
+            return (factor * yres[1] / yres[0], factor * xres[1] / xres[0])
```

Comment:
could you convert this type of logic into utility functions so that we can reuse?

Reply:
I implemented a utility class to convert units with a bit of generalization to other quantities. Please let me know what do you think?